### PR TITLE
Remove CompositeBuilder

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR is part 3 of a series where usages of the Match class will slowly be replaced by the underlying Predicate.

It might be easier to review the commits individually, no functional changes

### CODE REVIEW

File: src/main/java/games/strategy/triplea/ui/MovePanel.java [hunk 1]
```diff
@@ -351,35 +352,31 @@ private void sortTransportsToUnload(final List<Unit> transports, final Route rou
...
-      final Match.CompositeBuilder<Unit> rightUnitTypeMatchBuilder = Match.newCompositeBuilder();
-      for (final Unit unit : units) {
-        if (unit.getOwner().equals(owner)) {
-          rightUnitTypeMatchBuilder.add(Matches.unitIsOfType(unit.getType()));
-        }
-      }
-      movableBuilder.add(rightUnitTypeMatchBuilder.any());
+      movable = movable.and(u -> units.stream()
+          .filter(unit -> unit.getOwner().equals(owner))
+          .anyMatch(unit -> Matches.unitIsOfType(unit.getType()).test(u)));
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/triplea/ui/MovePanel.java]: I extracted this bit of code to a seperate method.
The name of this method makes the predicate much clearer.
Then I noticed how overcomplicated I made this method, so I replaced Matches.unitIsOfType with its implementation which boosts readability by a lot.

[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/triplea/ui/MovePanel.java]: Aaand then I noticed that I could indeed use the Matches.unitIsOfType method more easily which reuses code and sqashed all commits into a single one

File: src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java [hunk 2]
```diff
@@ -525,17 +526,16 @@ private static Route getMaxSeaRoute(final GameData data, final Territory start,
...
-      unitCondBuilder.add(sub);
-    }
+    final Predicate<Unit> unitCond = Matches.unitIsInfrastructure().invert()
+        .and(Matches.alliedUnit(player, data).invert())
+        .and(Properties.getIgnoreTransportInMovement(data)
+            ? transport
+            : Matches.always())
+        .and(Properties.getIgnoreSubInMovement(data)
+            ? sub
+            : Matches.always());
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java]: Not really a big fan of how this syntax ends up reading.

**Comment by ron-murhammer** [comment]
I quickly glanced through this. The one thing that I don't like about it is how it added lots of "always" and "never". Those essentially have no value and make things harder to read. Not sure if there is an easy way to avoid that.

File: src/main/java/games/strategy/triplea/ui/MovePanel.java [hunk 3]
```diff
@@ -351,35 +352,35 @@ private void sortTransportsToUnload(final List<Unit> transports, final Route rou
...
-        }
-      }
-      movableBuilder.add(rightUnitTypeMatchBuilder.any());
+      movable = movable.and(areOwnedUnitsOfType(units, owner));
     }
-    return movableBuilder.all();
+    return Match.of(movable);
+  }
+
+  private Predicate<Unit> areOwnedUnitsOfType(final Collection<Unit> units, final PlayerID owner) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: This method can be `static`.

File: src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java (hunk 2 above)
Comment (ssoloff) [reply src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java]: Yeah, this is basically a competition between the forces of readability and immutability.

One way to keep both forces happy is if there were an `and()` overload that accepted a Boolean condition and another predicate so that it essentially hid the use of `always()`.  However, since Java doesn't have something similar to .NET's extension methods, and we want to be able to keep chaining the method calls, adding such an overload would require us to keep the `Match<T>` class, whose total removal is @RoiEXLab's goal, I believe.

Just for kicks, such a method would look like:

```java
public final class Match<T> implements Predicate<T> {
  // ...

  @Override
  public Match<T> and(final Predicate<? super T> other) {
    return Match.of(Predicate.super.and(other));
  }

  public Match<T> andIf(final boolean condition, final Predicate<? super T> other) {
    return this.and(condition ? other : Matches.always());
  }
}
```

so that the tagged example above would now read:

```java
    final Match<Unit> unitCond = Matches.unitIsInfrastructure().invert()
        .and(Matches.alliedUnit(player, data).invert())
        .andIf(Properties.getIgnoreTransportInMovement(data), transport)
        .andIf(Properties.getIgnoreSubInMovement(data), sub);
```

The `andIf()` implementation can probably be simplified to return `this` in the case the condition is `false`:

```java
  public Match<T> andIf(final boolean condition, final Predicate<? super T> other) {
    return condition ? this.and(other) : this;
  }
```

Note that none of this was verified and, if adopted, unit tests for `andIf()` should be added. :smile:

Comment (ssoloff) [reply src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java]: Another alternative, if keeping `Match<T>` is not acceptable, is to introduce something like a `PredicateBuilder` class that has the required methods.  It would be a little more verbose, though:

```java
final Predicate<Unit> unitCond = PredicateBuilder.of(Matches.unitIsInfrastructure().invert())
    .and(Matches.alliedUnit(player, data).invert())
    .andIf(Properties.getIgnoreTransportInMovement(data), transport)
    .andIf(Properties.getIgnoreSubInMovement(data), sub)
    .build();
```

**Comment by RoiEXLab** [comment]
@ssoloff I'm probably going with your second suggestion, will take another day though

**Comment by RoiEXLab** [comment]
@ssoloff @ron-murhammer I added a PredicateBuilder to make the code more readable.

File: src/main/java/games/strategy/triplea/delegate/Matches.java [hunk 4]
```diff
@@ -2408,41 +2405,30 @@ private Matches() {}
...
-          canBeInBattle.add(unitTypeIsLand().invert());
-        }
-      } else { // defense
-        if (isLandBattle) {
-          canBeInBattle.add(unitTypeIsSea().invert());
-        } else { // is sea battle
-          canBeInBattle.add(unitTypeIsLand().invert());
+
+    // Filter out anything like factories, or units that have no combat ability AND cannot be taken casualty
+    final PredicateBuilder<UnitType> canBeInBattle = PredicateBuilder.of(unitTypeIsInfrastructure().invert())
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/delegate/Matches.java]: Nit: Might want to rename this `canBeInBattleBuilder`.  I had to take a double-take below when I saw statements like:

```java
canBeInBattle
    .and(unitTypeCanNotMoveDuringCombatMove().invert())
    .and(unitTypeCanMove(player));
```

and I didn't see the result being assigned back to `canBeInBattle` (i.e. I confused it for a `Predicate` instead of a `PredicateBuilder`).

File: src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java [hunk 5]
```diff
@@ -589,13 +587,12 @@ private static Route getMaxSeaRoute(final GameData data, final Territory start,
...
     // feel free to change, if you are confortable all calls to this function conform.
-    final Match.CompositeBuilder<Territory> endCondBuilder = Match.newCompositeBuilder(
-        Matches.territoryIsImpassable().invert());
-    if (Properties.getNeutralsImpassable(data)) {
-      endCondBuilder.add(Matches.territoryIsNeutralButNotWater().invert());
-    }
+    final Predicate<Territory> endCond = Matches.territoryIsImpassable().invert()
+        .and(Properties.getNeutralsImpassable(data)
+            ? Matches.territoryIsNeutralButNotWater().invert()
+            : Matches.always());
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java]: Could this be changed to use new builder methods?

File: src/main/java/games/strategy/triplea/ui/MovePanel.java [hunk 6]
```diff
@@ -803,19 +805,19 @@ private void selectUnitsToMove(final List<Unit> units, final Territory t, final
       // add all
       if (me.isShiftDown()) {
         // prevent units of multiple owners from being chosen in edit mode
-        final Match.CompositeBuilder<Unit> ownedNotFactoryBuilder = Match.newCompositeBuilder();
+        final PredicateBuilder<Unit> ownedNotFactory = PredicateBuilder.of(Matches.always());
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: Nit: Could use a `Builder` suffix as `movableBuilder` above.

File: src/main/java/games/strategy/util/PredicateBuilder.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package games.strategy.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.function.Predicate;;
+
+public class PredicateBuilder<T> {
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/PredicateBuilder.java]: Needs a Javadoc to make Checkstyle happy.

File: src/main/java/games/strategy/util/PredicateBuilder.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
+package games.strategy.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.function.Predicate;;
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/PredicateBuilder.java]: Extra semi-colon is making Checkstyle angry.

File: src/main/java/games/strategy/triplea/ui/MovePanel.java [hunk 9]
```diff
@@ -327,19 +329,19 @@ private void sortTransportsToUnload(final List<Unit> transports, final Route rou
   }
 
   private Match<Unit> getMovableMatch(final Route route, final Collection<Unit> units) {
-    final Match.CompositeBuilder<Unit> movableBuilder = Match.newCompositeBuilder();
+    final PredicateBuilder<Unit> movableBuilder = PredicateBuilder.of(Matches.always());
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: Would it be worth having an initial builder method that creates a "true" predicate? Something like PredicateBuilder.always() or PredicateBuilder.true()?

File: src/main/java/games/strategy/triplea/ui/MovePanel.java (hunk 6 above)
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: Would it be worth having an initial builder method that creates a "true" predicate? Something like PredicateBuilder.always() or PredicateBuilder.true()?

File: src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java (hunk 5 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java]: I almost expected to miss something... will fix

File: src/main/java/games/strategy/util/PredicateBuilder.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+   */
+  public static <T> PredicateBuilder<T> of(final Predicate<T> predicate) {
+    return new PredicateBuilder<>(predicate);
+  }
+
+  /**
+   * Modifies this PredicateBuilder to return the underlying predicate
+   * combined with a logical AND with the given Predicate.
+   */
+  public PredicateBuilder<T> and(final Predicate<T> predicate) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/PredicateBuilder.java]: This may be a YAGNI, but I'm not sure if the parameters to `and()`, `andIf()`, `or()`, and `orIf()` should be `Predicate<? super T>` to match the capabilities of `Predicate#and()` and `Predicate#or()`.

Nit: Also, renaming the parameters to `other`, as the `Predicate` methods do, would avoid having to qualify `this.predicate` where it is currently necessary.

**Comment by RoiEXLab** [comment]
@ssoloff @ron-murhammer Updated!

File: src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java [hunk 11]
```diff
@@ -256,15 +257,10 @@ public static void incrementUnitProductionForBidTerritories(
...
-        Matches.unitIsBeingTransported().invert());
-    if (territory.isWater()) {
-      factoryMatchBuilder.add(Matches.unitIsLand().invert());
-    } else {
-      factoryMatchBuilder.add(Matches.unitIsSea().invert());
-    }
-    final Collection<Unit> factoryUnits = territory.getUnits().getMatches(factoryMatchBuilder.all());
+    final Predicate<Unit> factoryMatch = Matches.unitIsOwnedAndIsFactoryOrCanProduceUnits(player)
+        .and(Matches.unitIsBeingTransported().invert())
+        .and((territory.isWater() ? Matches.unitIsLand() : Matches.unitIsSea()).invert());
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java]: I think you are missing an invert() on isLand here.

File: src/main/java/games/strategy/triplea/delegate/Matches.java [hunk 12]
```diff
@@ -2408,41 +2405,30 @@ private Matches() {}
...
+      }
+      if (isLandBattle) {
+        if (doNotIncludeBombardingSeaUnits) {
+          canBeInBattleBuilder.and(unitTypeIsSea().invert());
         }
+      } else { // is sea battle
+        canBeInBattleBuilder.and(unitTypeIsLand().invert());
       }
+    } else { // defense
+      canBeInBattleBuilder.and((isLandBattle ? unitTypeIsSea() : unitTypeIsLand()).invert());
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/triplea/delegate/Matches.java]: Missing invert on unitTypeIsSea()?

@ssoloff @RoiEXLab Can you both double check that this method is equivalent after the refactoring as the diff is hard to follow.

File: src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java (hunk 11 above)
Comment (ssoloff) [reply src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java]: Yeah, this tripped me up as well during the first review.  There's parentheses wrapping the entire ternary expression, so that `invert()` is applied to the value of that expression, not only to `Matches.unitIsSea()`.

File: src/main/java/games/strategy/triplea/delegate/Matches.java (hunk 12 above)
Comment (ssoloff) [reply src/main/java/games/strategy/triplea/delegate/Matches.java]: Same as above.

**Comment by RoiEXLab** [comment]
@ssoloff Already explained this, I'm going ahead an merge this, in case you want the notation changed I'd be happy to do that in a follow-up PR.

File: src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java (hunk 11 above)
Comment (ron-murhammer) [reply src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java]: Ah. I see now. Missed the parenthesis.
